x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java

```diff
@@ -0,0 +1,491 @@
+public final class StringRuntimeFieldHelper {
+    @FunctionalInterface
+    public interface NewLeafLoader {
+        IntConsumer leafLoader(LeafReaderContext ctx, Consumer<String> sync) throws IOException;
+    }
+
+    private final NewLeafLoader newLeafLoader;
+
+    public StringRuntimeFieldHelper(NewLeafLoader newLeafLoader) {
+        this.newLeafLoader = newLeafLoader;
+    }
+
+    public CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+        return new Values().docValues();
+    }
+
+    public Query existsQuery(String fieldName) {
+        return new Values().new ExistsQuery(fieldName);
+    }
+
+    public Query fuzzyQuery(String fieldName, String value, int maxEdits, int prefixLength, int maxExpansions, boolean transpositions) {
+        return new Values().new FuzzyQuery(fieldName, value, maxEdits, prefixLength, maxExpansions, transpositions);
+    }
+
+    public Query prefixQuery(String fieldName, String value) {
+        return new Values().new PrefixQuery(fieldName, value);
+    }
+
+    public Query rangeQuery(String fieldName, String lowerValue, String upperValue, boolean includeLower, boolean includeUpper) {
+        return new Values().new RangeQuery(fieldName, lowerValue, upperValue, includeLower, includeUpper);
+    }
+
+    public Query regexpQuery(String fieldName, String pattern, int flags, int maxDeterminizedStates) {
+        return new Values().new RegexpQuery(fieldName, pattern, flags, maxDeterminizedStates);
+    }
+
+    public Query termQuery(String fieldName, String value) {
+        return new Values().new TermQuery(fieldName, value);
+    }
+
+    public Query termsQuery(String fieldName, String... values) {
+        return new Values().new TermsQuery(fieldName, values);
+    }
+
+    public Query wildcardQuery(String fieldName, String pattern) {
+        return new Values().new WildcardQuery(fieldName, pattern);
+    }
+
+    private class Values extends AbstractRuntimeValues {
+        private String[] values = new String[1];
+
+        @Override
+        protected IntConsumer newLeafLoader(LeafReaderContext ctx) throws IOException {
+            return newLeafLoader.leafLoader(ctx, this::add);
+        }
+
+        private void add(String value) {
+            if (value == null) {
+                return;
+            }
+            int newCount = count + 1;
+            if (values.length < newCount) {
+                values = Arrays.copyOf(values, ArrayUtil.oversize(newCount, RamUsageEstimator.NUM_BYTES_OBJECT_REF));
+            }
+            values[count] = value;
+            count = newCount;
+        }
+
+        @Override
+        protected void sort() {
+            Arrays.sort(values, 0, count);
+        }
+
+        private CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+            alwaysSortResults();
+            return DocValues::new;
+        }
+
+        private class DocValues extends SortedBinaryDocValues {
```

Comment:
like I said in other comments, I think that you have all the bits that are needed, I only would turn around a bit the dependencies/flow. I would keep the existing doc_values impl, which is only used in aggs/sorting (and it is totally fine that it's not used in queries), and I would make it lighter so that all it does is really returning the result of the script for the current doc, by calling that script.getResult(docId) method I mentioned above. I think that would clarify who does what. At the moment it is hard to follow who needs doc_values who needs them sorted etc.